### ISSUE TITLE
[FEATURE] Allow users to set count/size limit on snapshots

### ISSUE THREAD

**Issue by FrankYang0529** [harvester/harvester#4478]
**Is your feature request related to a problem? Please describe.**
Snapshots use extra storage. Currently, users don't have a way to control snapshots count/size and check how much size snapshot uses. Snapshots may use more storage than users expect.

**Describe the solution you'd like**
Leverage [LH snapshot space management](https://github.com/longhorn/longhorn/issues/6563) feature to achieve it.

**Comment by tserong** [comment]
Does this need a "require/doc" label? Also, do we need GUI changes for this feature, or are users expected to set the annotations up by editing the annotations in the volume YAML?

**Comment by m-ildefons** [comment]
Needs documentation update.
It's unclear what units are expected, e.g.:
`.spec.snapshotLimit.namespaceTotalSnapshotSizeQuota: 100` is this 100 snapshots, 100 bytes or 100 GiB?

**Comment by TachunLin** [comment]
Verified **fixed** on Harvester `v1.4.0-rc3` with Rancher `v2.9.2`. Close this issue

### Result

<details><summary>$\color{green}{\textsf{PASS}}$ Check <strong>namespaceTotalSnapshotSizeUsage</strong> and <strong>vmTotalSnapshotSizeUsage</strong> values format $~~$</summary>

1. Can't set `namespaceTotalSnapshotSizeUsage` to negative value 
  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # resourcequotas.harvesterhci.io "default-resource-quota" was not valid:
  # * spec.snapshotLimit.namespaceTotalSnapshotSizeQuota: Invalid value: -1: spec.snapshotLimit.namespaceTotalSnapshotSizeQuota in body should be greater than or equal to 0
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ResourceQuota
  ...
  spec:
    snapshotLimit:
      namespaceTotalSnapshotSizeQuota: -1
      vmTotalSnapshotSizeQuota:
        vm: 32212254720
  ```

2. Can't set `vmTotalSnapshotSizeUsage` to negative value
  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # resourcequotas.harvesterhci.io "default-resource-quota" was not valid:
  # * spec.snapshotLimit.vmTotalSnapshotSizeQuota[vm]: must be greater than or equal to 0
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ResourceQuota
  metadata:
    creationTimestamp: "2024-10-21T06:11:04Z"
    generation: 8
    name: default-resource-quota
    namespace: default
    resourceVersion: "2716201"
    uid: 16601199-dc77-4739-8c53-be0ed4d0e215
  spec:
    snapshotLimit:
      namespaceTotalSnapshotSizeQuota: 32212254720
      vmTotalSnapshotSizeQuota:
        vm: -1
  ```

3. Can't set `namespaceTotalSnapshotSizeUsage` to non-integer value 

  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # resourcequotas.harvesterhci.io "default-resource-quota" was not valid:
  # * spec.snapshotLimit.namespaceTotalSnapshotSizeQuota: Invalid value: "string": spec.snapshotLimit.namespaceTotalSnapshotSizeQuota in body must be of type integer: "string"
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ResourceQuota
  metadata:
    creationTimestamp: "2024-10-21T06:11:04Z"
    generation: 8
    name: default-resource-quota
    namespace: default
    resourceVersion: "2716201"
    uid: 16601199-dc77-4739-8c53-be0ed4d0e215
  spec:
    snapshotLimit:
      namespaceTotalSnapshotSizeQuota: test
      vmTotalSnapshotSizeQuota:
        vm: 32212254720
  ```


4. Can't set `vmTotalSnapshotSizeUsage` to non-integer value 
  ```
  # Please edit the object below. Lines beginning with a '#' will be ignored,
  # and an empty file will abort the edit. If an error occurs while saving this file will be
  # reopened with the relevant failures.
  #
  # resourcequotas.harvesterhci.io "default-resource-quota" was not valid:
  # * spec.snapshotLimit.vmTotalSnapshotSizeQuota.vm: Invalid value: "string": spec.snapshotLimit.vmTotalSnapshotSizeQuota.vm in body must be of type integer: "string"
  #
  apiVersion: harvesterhci.io/v1beta1
  kind: ResourceQuota
  metadata:
    creationTimestamp: "2024-10-21T06:11:04Z"
    generation: 8
    name: default-resource-quota
    namespace: default
    resourceVersion: "2716201"
    uid: 16601199-dc77-4739-8c53-be0ed4d0e215
  spec:
    snapshotLimit:
      namespaceTotalSnapshotSizeQuota: 32212254720
      vmTotalSnapshotSizeQuota:
        vm: test
  ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test the <strong>vmTotalSnapshotSizeUsage</strong> setting on VM $~~$</summary>


1. Write data to make two volumes full. (10Gi and 20Gi)
  ```
  Filesystem      Size  Used Avail Use% Mounted on
  devtmpfs        4.0M  8.0K  4.0M   1% /dev
  tmpfs           979M  4.0K  979M   1% /dev/shm
  tmpfs           392M  5.9M  386M   2% /run
  tmpfs           4.0M     0  4.0M   0% /sys/fs/cgroup
  /dev/vda3        10G   10G   36K 100% /
  /dev/vda2        20M  3.0M   17M  15% /boot/efi
  tmpfs           196M     0  196M   0% /run/user/1000
  /dev/vdb         20G   20G     0 100% /mnt/vdb
  ```

2.  ✅ Create the first VMBackup. It should be allowed. The total snapshot size is 30Gi.  

  ![image](https://github.com/user-attachments/assets/370cfca1-a566-4b86-b235-d02fa604c432)
  
  ![image](https://github.com/user-attachments/assets/8168d7fe-38e3-4782-affb-b597cfb6a3be)


3.  Clear data and write data to make two volumes full again 

  ```
  vm:~ # df -h
  Filesystem      Size  Used Avail Use% Mounted on
  devtmpfs        4.0M  8.0K  4.0M   1% /dev
  tmpfs           979M  4.0K  979M   1% /dev/shm
  tmpfs           392M  5.9M  386M   2% /run
  tmpfs           4.0M     0  4.0M   0% /sys/fs/cgroup
  /dev/vda3        10G   10G   36K 100% /
  /dev/vda2        20M  3.0M   17M  15% /boot/efi
  tmpfs           196M     0  196M   0% /run/user/1000
  /dev/vdb         20G   20G     0 100% /mnt/vdb
  ```

4. Create another VM Snapshot. It should be allowed. The total snapshot size is 60Gi

![image](https://github.com/user-attachments/assets/37f26384-27aa-4d66-be70-620655538cc7)

  ```
  NAMESPACE         NAME                                            VOLUME                                     CREATIONTIME           READYTOUSE   RESTORESIZE   SIZE          AGE
  longhorn-system   snapshot-15bfa0a4-8327-49ef-bca6-9f653a7fb092   pvc-1ef5d807-1aab-46bc-9442-fa2642a571c1   2024-10-21T08:44:49Z   true         10737418240   8599547904    97s
  longhorn-system   snapshot-8bd5a4fe-84fe-424b-9afb-477d6e58a5e6   pvc-157c4541-db9c-4e39-a391-54f4d8f6a79a   2024-10-21T08:32:41Z   true         21474836480   21415469056   13m
  longhorn-system   snapshot-c0ca8527-447e-4c36-b343-d043661c76b9   pvc-1ef5d807-1aab-46bc-9442-fa2642a571c1   2024-10-21T08:32:41Z   true         10737418240   8974888960    13m
  longhorn-system   snapshot-e350c427-26af-4b10-8418-d43a1749b0e3   pvc-157c4541-db9c-4e39-a391-54f4d8f6a79a   2024-10-21T08:44:50Z   true         21474836480   20943081472   96
  ```

5. ✅ Set 30Gi virtual machine quota on the VM, It should be allowed

  ![image](https://github.com/user-attachments/assets/bb02a1bd-4dda-4f90-8f83-5b879a1e64d1)
  
  ![image](https://github.com/user-attachments/assets/63b86ce4-57a9-4b91-ada8-7cc800bb91fe)


6. Clear data and write data to make two volumes full again

7. ✅ Create the second VMBackup. It should be `denied`, because the total snapshot size will be 90Gi

  ```
  failed to create VM backup, error: admission webhook "validator.harvesterhci.io" denied the request: total snapshot size limit 30Gi is less than total snapshot usage 87378588Ki
  ```
  
  ![image](https://github.com/user-attachments/assets/59bfe21b-28bb-4ba6-9db9-8854ea06e935)

8. ✅ Remove the first backup and the first snapshot. The total snapshot size is `0Gi`

![image](https://github.com/user-attachments/assets/db323817-1c03-47bf-bd45-f3ca829a7cc1)
 
9. ✅ Create a new VMBackup. It should be allowed. The total snapshot size is 30Gi 

  ![image](https://github.com/user-attachments/assets/dc60b6a7-aab7-4db2-940f-5ef40b9f198d)
  
  ![image](https://github.com/user-attachments/assets/222a6320-bbff-4b0b-9b8d-aedc5b0572d4)
  
  ![image](https://github.com/user-attachments/assets/8854fe51-b314-4af2-9cac-ff05f003bbe1)


 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Test the <strong>namespaceTotalSnapshotSizeQuota</strong> setting on namespace for VM $~~$</summary>

1. Create a VM with size `10Gi`
2. ✅ Can set `10Gi` snapshot size quota on a Namespace
  ![image](https://github.com/user-attachments/assets/cb8575e0-c552-4a7c-ae44-fabf05a2eee7)

3. Write data to make only volumes full. (10Gi) 

  ```
  opensuse@vm1:~> df -h
  Filesystem      Size  Used Avail Use% Mounted on
  devtmpfs        4.0M  8.0K  4.0M   1% /dev
  tmpfs           979M  4.0K  979M   1% /dev/shm
  tmpfs           392M  5.9M  386M   2% /run
  tmpfs           4.0M     0  4.0M   0% /sys/fs/cgroup
  /dev/vda3        10G   10G   40K 100% /
  /dev/vda2        20M  3.0M   17M  15% /boot/efi
  tmpfs           196M     0  196M   0% /run/user/1000
  ```

2. ✅ Create the first VMBackup. It should be allowed. The total snapshot size is `10Gi`

  ![image](https://github.com/user-attachments/assets/4c7f1a52-6349-4272-b68c-0d13e47b18e9)

  ![image](https://github.com/user-attachments/assets/07107737-641b-48df-8d8e-1051b7a8edcb)


3. Clear data and write data to make the volume full again. 

4. ✅ Create the second VMBackup. It should be denied, because the total snapshot size will be `20Gi`

  ```
  failed to create VM backup, error: admission webhook "validator.harvesterhci.io" denied the request: total snapshot size limit 10Gi is less than total snapshot usage 17162808Ki
  ```

  ![image](https://github.com/user-attachments/assets/b447353e-6121-4a16-9421-09ed3d522e0a)


5. Remove the first VMBackup.

  ![image](https://github.com/user-attachments/assets/debd1bda-71e4-4ee6-b56f-7886af4ebbaf)


6. Create a new VMBackup. It should be allowed. The total snapshot size is `10Gi`

  ![image](https://github.com/user-attachments/assets/11de5d3c-aac1-4109-9d5f-0faa61445cab)
  
  ![image](https://github.com/user-attachments/assets/10b16686-b428-4cba-85ac-99cca93af354)

  ![image](https://github.com/user-attachments/assets/e44ebf79-b263-4b05-a2f8-658d4728cb9e)


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test the <strong>namespaceTotalSnapshotSizeQuota</strong> setting on namespace for Volume $~~$</summary>

1. Create a new VM with size `10Gi`

2. Set `10Gi` snapshot size quota on a Namespace

3. Write data to make only volumes full. (10Gi) 



2. ✅ Create the first Volume Snapshot. It should be allowed. The total snapshot size is `10Gi`


  ![image](https://github.com/user-attachments/assets/f464cae0-01f4-4a01-9407-9f8d940c8ee4)
  ![image](https://github.com/user-attachments/assets/195281ee-bde3-47dd-bb10-cc1f373f7683)


3. Clear data and write data to make the volume full again. 

4. ✅ Create the second Volume Snapshot. It should be denied, because the total snapshot size will be `20Gi`

  ```
  admission webhook "validator.harvesterhci.io" denied the request: total snapshot size limit 10Gi is less than total snapshot usage 17230224Ki
  ```
  ![image](https://github.com/user-attachments/assets/13511b74-57f4-425f-a6e4-4cca2531ad09)



5. Remove the first Volume Snapshot.


6. Create a new Volume Snapshot. It should be allowed. The total snapshot size is `10Gi`

  ![image](https://github.com/user-attachments/assets/d1b0ddbb-12ad-4ac4-8102-2ee361f97d87)
  
  ![image](https://github.com/user-attachments/assets/16c0009b-3970-4d9f-8f40-c68dfa88d6f3)



 </details>


<details><summary>$\color{green}{\textsf{PASS}}$ Test users with <strong>Harvester ResourceQuotas Read Only </strong> role $~~$</summary>

1. Apply `harvester-resourcequotas-admin` and `harvester-resourcequotas-read-only` to a Rancher.

2. Create two standard user in Rancher Users page
  ![image](https://github.com/user-attachments/assets/fded7e9e-782b-4e91-a235-8fc65c787aeb)

3. Create a new user with `Harvester ResourceQuotas Read Only` role in Harvester RBAC 

  ![image](https://github.com/user-attachments/assets/0f06d27d-9334-44ea-9590-d7a82aebc19f)
  ![image](https://github.com/user-attachments/assets/1ab9c9ff-7d87-4bf2-8607-c0bbad42eb33)


4.  Add the user to the `default namespace project as member.

  ![image](https://github.com/user-attachments/assets/cda647d3-69d3-4c01-b6b5-fc896803ec0a)


5. With read only quota permission user `can't` update the snapshot quota on **namespace**, return `403` code message 

 - Using post man
  ![image](https://github.com/user-attachments/assets/1af420f7-3263-47c8-ad4e-be91ff437ad5)

 - Using curl command
  ```
  PermissionDenied 403: User does not have permission to update resource quota
  ```

6. With read only quota permission user `can't` update the snapshot quota on **VM**, return `403` code message 

 - Using post man

  ![image](https://github.com/user-attachments/assets/c62457e1-b90d-4591-8cbd-fa85bcc6d96b)


 - Using curl command

  ```
  PermissionDenied 403: User does not have permission to update resource quota
  ```


 </details>

<details><summary>$\color{green}{\textsf{PASS}}$ Test users with <strong>Harvester ResourceQuotas Admin</strong> role $~~$</summary>

1. Apply `harvester-resourcequotas-admin` and `harvester-resourcequotas-read-only` to a Rancher.

2. Create two standard user in Rancher Users page
  ![image](https://github.com/user-attachments/assets/fded7e9e-782b-4e91-a235-8fc65c787aeb)

3. Create a new user with `Harvester ResourceQuotas Admin` role in Harvester RBAC 

  ![image](https://github.com/user-attachments/assets/8b072847-4b40-497f-a382-46d1a904bd10)

4.  Add the user to the `default namespace project as member.

  ![image](https://github.com/user-attachments/assets/b3b939c5-0eac-4f01-9e32-0749df25b81d)


5. With admin quota permission user `can` update the snapshot quota on **namespace**
 - Using post man

  ![image](https://github.com/user-attachments/assets/6f006acd-c68b-4156-a975-b074660e53f0)
  ![image](https://github.com/user-attachments/assets/8a65debc-f5cc-4567-b415-d0b6a50d2e81)

6. With admin quota permission user `can` update the snapshot quota on **VM** 

 - Using post man

  ![image](https://github.com/user-attachments/assets/2c3545f1-9ff6-4517-82b8-955c018404b4)
  
  ![image](https://github.com/user-attachments/assets/a7be11d4-8ee1-42ca-b375-783fce0bb21c)


 </details>

### Test Information
* Test Environment: Single node harvester on baremetal machine
* Harvester version:  `v1.4.0-rc3`
* Rancher version: `v2.9.2

### Verify Steps
<details><summary>Check <strong>namespaceTotalSnapshotSizeUsage</strong> and <strong>vmTotalSnapshotSizeUsage</strong> values format </summary>

1. Click the `Edit Virtual Machine Quota` setting on the VM 
8. Set the value to Edit Quota
9. Click the `Edit Quota` setting on the Namespace
10. Set the value to Edit Quota
11. Open K9s
12. Search for the `resourcequotas.harvesterhci.io` CR
13. Edit the CR
14. Set the `namespaceTotalSnapshotSizeUsage` to negative and non-integer
15. Set the `vmTotalSnapshotSizeUsage` to negative and non-integer

</details>

<details><summary>Test the <strong>vmTotalSnapshotSizeUsage</strong> setting on VM </summary>

1. Clear all the setting of `namespaceTotalSnapshotSizeUsage` and `vmTotalSnapshotSizeUsage`
2. Create a VM has two volumes. The first volume size is 10Gi and the second volume size is 20Gi.
3. Write data to make two volumes full. 
```
sudo mkfs.ext4 /dev/vdb
sudo mkdir /mnt/vdb
mount /dev/vdb /mnt/vdb

sudo dd if=/dev/zero of=/bigfile bs=1M
sudo dd if=/dev/zero of=/bigfile bs=1M
```
6. Create the first VMBackup. It should be allowed. The total snapshot size is 30Gi.
7. Clear data and write data to make two volumes full again
8. Create the first VM Snapshot. It should be allowed, The total snapshot size is 60Gi.
9. Set 30Gi on the VM snapshot quota. It should be allowed.
10. Clear data and write data to make two volumes full again.
11. Create the second VMBackup. It should be denied, because the total snapshot size will be 90Gi if the request is allowed.
16. Remove the first and the second VMBackup. The total snapshot size is 0Gi
17. Create a new VMBackup. It should be allowed.
18. Check the total snapshot size is 30Gi


</details>

<details><summary>Test the <strong>namespaceTotalSnapshotSizeQuota</strong> setting on namespace for VM </summary>

1. Clear all the setting of `namespaceTotalSnapshotSizeUsage` and `vmTotalSnapshotSizeUsage`
2. Create a VM has one volume with volume size is 10Gi
3. Set `10Gi` snapshot size quota on a Namespace
4. Write data to make two volumes full. 
```
sudo dd if=/dev/zero of=/bigfile bs=1M
```
5. Create the first VMBackup. It should be allowed. The total snapshot size is 10Gi.
6. Clear data and write data to make the volume full again. 
7. Create the second VMBackup. It should be denied,
8. Remove the first VMBackup. The total snapshot size is 0Gi.
9. Create a new VMBackup. It should be allowed, The total snapshot size is 10Gi


</details>

<details><summary>Test the <strong>namespaceTotalSnapshotSizeQuota</strong> setting on namespace for Volume </summary>

1. Clear all the setting of `namespaceTotalSnapshotSizeUsage` and `vmTotalSnapshotSizeUsage`
2. Create a VM has one volume with volume size is 10Gi
3. Set `10Gi` snapshot size quota on a Namespace
4. Write data to make two volumes full. 
```
sudo dd if=/dev/zero of=/bigfile bs=1M
```
5. Go to the volumes page, find the 10Gi volume of the vm
6. Create the first volume snapshot. It should be allowed. The total snapshot size is 10Gi.
7. Clear data and write data to make the volume full again. 
8. Create the second volume snapshot. It should be denied,
9. Remove the first volume snapshot. The total snapshot size is 0Gi.
12. Create a new volume snapshot. It should be allowed, The total snapshot size is 10Gi

</details>


<details><summary>Test users with <strong>Harvester ResourceQuotas Read Only </strong> role </summary>

1. Apply harvester-resourcequotas-admin and harvester-resourcequotas-read-only to a Rancher
2. Import Harvester to a Rancher.
3. Create a standard user named `quota_readonly` in Rancher
4. Access Harvester from Rancher 
5. Open the RBAC page 
6. Create a new user with Harvester ResourceQuotas Read Only role.
7. Open the names space page
8. Add the user to the `default` namespace project as member.
9. Logout the Rancher admin account
10. Login with the `quota_readonly` user
11. Open the Preference -> Account and API Keys
12. Create API key for this user
13. Get the token of the `Bearer Token`
14. Inject the token in to the API request
15. Send requests to update Resource Quotas via API on Namespace, requests should be denied.

  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer token-xxxx" \
       -d '{
            "totalSnapshotSizeQuota": "70Gi"
           }' \
       https://<Rancher IP>/k8s/clusters/<cluster id>/v1/harvester/namespace/default?action=updateResourceQuota -k
  ```

16. Send requests to update Resource Quotas via API on VM, requests should be denied.

  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer token-xxx" \
       -d '{
            "totalSnapshotSizeQuota": "80Gi"
           }' \
       https://<Rancher IP>/k8s/clusters/<cluster id>/v1/harvester/kubevirt.io.virtualmachines/default/vm1?action=updateResourceQuota -k
  ```


</details>


<details><summary>Test users with <strong>Harvester ResourceQuotas Admin</strong> role </summary>

1. Apply harvester-resourcequotas-admin and harvester-resourcequotas-read-only to a Rancher
2. Import Harvester to a Rancher.
3. Create a standard user named `quota_admin` in Rancher
4. Access Harvester from Rancher 
5. Open the RBAC page 
6. Create a new user with Harvester ResourceQuotas Admin role.
7. Open the names space page
8. Add the user to the `default` namespace project as member.
9. Logout the Rancher admin account
10. Login with the `quota_admin` user
11. Open the Preference -> Account and API Keys
12. Create API key for this user
13. Get the token of the `Bearer Token`
14. Inject the token in to the API request
15. Send requests to update Resource Quotas via API on Namespace, requests should be accepted

  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer token-xxxx" \
       -d '{
            "totalSnapshotSizeQuota": "70Gi"
           }' \
       https://<Rancher IP>/k8s/clusters/<cluster id>/v1/harvester/namespace/default?action=updateResourceQuota -k
  ```

16. Check the snapshot size quota value on namespace page

17. Send requests to update Resource Quotas via API on VM, requests should be accepted

  ```
  curl -X POST -H "Content-Type: application/json" -H "Authorization: Bearer token-xxx" \
       -d '{
            "totalSnapshotSizeQuota": "80Gi"
           }' \
       https://<Rancher IP>/k8s/clusters/<cluster id>/v1/harvester/kubevirt.io.virtualmachines/default/vm1?action=updateResourceQuota -k
  ```

18. Check the snapshot size quota value on `vm1` VM details page


</details>

**Comment by TachunLin** [comment]
During the issue verification. I notice the following behaviors worth mentioned for the reference. 

* All virtual machines under the same namespace would share with the snapshot size quota set on it.

* If no quota size limit set on the namespace, then each virtual machine would check and follow it's own quota size set on the vm level.
